plugins/org.jboss.tools.openshift.common.ui/src/org/jboss/tools/openshift/internal/common/ui/explorer/BaseExplorerContentProvider.java

```diff
@@ -182,121 +171,52 @@ protected IStatus run(IProgressMonitor monitor) {
 		job.schedule();
 		return stub;
 	}
-	
-	private void asyncViewerRefresh(final Runnable runner) {
-		Control control = viewer.getControl();
-		if(control.isDisposed()) {
-			return;
-		}
-		control.getDisplay().asyncExec(runner);
-	}
-	
-	protected void addChildrenToViewer(final Object parent, final Object ... objects) {
-		asyncViewerRefresh(new Runnable() {
-			@Override
-			public void run() {
-				synchronized (viewer) {
-					viewer.add(parent,objects);
-				}
-			}
-		});
-	}
-
-	protected void removeChildrenFromViewer(final Object parent, final Object ... objects) {
-		asyncViewerRefresh(new Runnable() {
-			@Override
-			public void run() {
-				synchronized (viewer) {
-					viewer.remove(parent, objects);
-				}
-			}
-		});
-	}
-	protected void updateChildrenFromViewer(final Object ... objects) {
-		asyncViewerRefresh(new Runnable() {
-			@Override
-			public void run() {
-				synchronized (viewer) {
-					viewer.update(objects, null);
-				}
-			}
-		});
-	}
 
 	protected void refreshViewer(final Object object) {
-		asyncViewerRefresh(new Runnable() {
-			@Override
-			public void run() {
-				synchronized (viewer) {
-					if(object != null) {
-						viewer.refresh(object);
-					}else {
-						viewer.refresh();
+		Control control = viewer.getControl();
+		if (!control.isDisposed()) {
+			control.getDisplay().asyncExec(new Runnable() {
+				@Override
+				public void run() {
+					if (!viewer.getControl().isDisposed()) {
```

Comment:
As a general rule, UI elements should not be manipulated outside the UI thread. The viewer variable is only ever changed on the UI thread. Which leaves stuff like "viewer.getControl()", etc. But again, I believe that these variables are constant after initialization, so no problem there.